### ISSUE TITLE
Support of linux hwmon (ina219)

### ISSUE THREAD

**Issue by jviki** [networkupstools/nut#2378]
Would it be possible to integrate support for standard linux hwmon that can monitor power of the system? I am running a small system (baseboard: [CM4-POE-UPS-BASE](http://www.waveshare.com/wiki/CM4-POE-UPS-BASE)) with batteries that contains the [ina219](https://docs.kernel.org/hwmon/ina2xx.html) I2C device. I hoped that I can find some generic hwmon UPS-like support here but I didn't.

I've got a short bash script that can be used as a basis for a nut driver:

```
#! /bin/bash

find_ina219()
{
	for d in $(ls /sys/class/hwmon/); do
		path="/sys/class/hwmon/${d}/name"
		name=$(cat ${path})

		if [ "${name}" == ina219 ]; then
			echo "/sys/class/hwmon/${d}"
			return 0
		fi
	done

	return 1
}

ina219=$(find_ina219)

if [ "$?" != 0 ]; then
	echo "could not find ina219 device" >&2
	exit 1
fi

voltage=`cat ${ina219}/in1_input`
current=`cat ${ina219}/curr1_input`

echo "Status: "$(expr \( ${voltage} - 3000 \) / 12)" % (${voltage} mV, ${current} mA)"

if [ "${voltage}" -lt 3150 ] && [ "${current}" -gt 50 ]; then 
	echo "Low power"
	sync
fi
```

Quick test:
```
# status: OL, HB
$ ./check-ups.sh 
Status: 95 % (4140 mV, 0 mA)

# status: OB, DISCHRG (unplugged AC)
$ ./check-ups.sh 
Status: 90 % (4084 mV, 87 mA)
$ ./check-ups.sh 
Status: 89 % (4072 mV, 114 mA)

# status: OL, CHRG (plugged AC back)
$ ./check-ups.sh 
Status: 102 % (4224 mV, -150 mA)
```

> if the current is positive, it means the output current of the battery (which can be considered as the Raspberry Pi load current), and if the current is negative, it means the battery charging current.
> -- [Battery Level Detection](https://www.waveshare.com/wiki/CM4-POE-UPS-BASE#INA219_Battery_Level_Detection)

Where should I start when trying to integrate this?

**Comment by jimklimov** [comment]
There are a few NUT drivers using the Linux i2c bus, a relatively recent one is pijuice. I suppose taking a look at those (and their integration to build system in makefiles) would be a helpful start.

**Comment by jimklimov** [comment]
I wonder if the new driver (or its future iterations) could also talk to devices like "Waveshare UPS HAT (E) for Raspberry Pi 5/4/3B+ that takes four 21700 Lithium batteries, supports USB PD 3.0"
* https://www.cnx-software.com/2024/08/03/waveshare-ups-hat-e-for-raspberry-pi-5-4-3b-takes-four-21700-lithium-batteries-supports-usb-pd-3-0/ 
* https://www.waveshare.com/wiki/UPS_HAT_(E)

**Comment by jviki** [comment]
It should, or at least I cannot see any reason why not... It's mostly about having the DTS right and loading the kernel driver.

I've got https://www.waveshare.com/wiki/UPS_HAT on the table and want to play with that as soon as I have some time.

**Comment by jimklimov** [comment]
Sounds great! If there are any more PRs to add support, would be great to get them :) And then perhaps let WaveShare know (or even ask for support via more devices to test against)

**Comment by mpn01** [comment]
> It should, or at least I cannot see any reason why not... It's mostly about having the DTS right and loading the kernel driver.
> 
> I've got https://www.waveshare.com/wiki/UPS_HAT on the table and want to play with that as soon as I have some time.

Hi, I'm also using UPS hat from _Waveshare_ and I was wondering if you managed to make it work? I have some issues with connecting it to _NUT_. 

I know it's a old thread but I haven't seen any one using this hat with _NUT_.

**Comment by jviki** [comment]
First, there is #2430, the module hwmon_ina219 was merged.

Unfortunately, I did not succeeded with the UPS HAT. It did not work out-of-the-box for me, I suspect there was some hardware issue. But, I didn't have time to figure out more details. Technically, it should be easy, the module should support it well. You just need to configure it properly.

**Comment by mpn01** [comment]
Thanks for your reply!

Sorry to hear that you didn't succeed.

After my comment I've kind of make it work but not fully. The Waveshare script worked, but system didn't see anything new in `/sys/class/hwmon/`. After some tinkering, I changed the overlay file (see comments).

```
/dts-v1/;
/plugin/;

/ {
    compatible = "brcm,bcm2835";

    fragment@0 {
        target = <&i2c_arm>;  # change here from &i2c_csi_dsi to &i2c_arm or &i2c_vc (don't remember correctly)
        __overlay__ {
            #address-cells = <1>;
            #size-cells = <0>;

            ina219@42 {
                status = "okay";
                compatible = "ti,ina219"; 
                reg = <0x42>; # change here from 0x43. for me it's 0x42
                shunt-resistor = <100000>;  // R100
            };
        };
    };
};
```
After converting, system properly showed new `hwmon` device. 

Then I started trying to connect it to NUT. But since Raspberry Pi OS is based on Debian 13, there is only `NUT 2.8.1` (and probably there won't be newer version in this release of Debian) which didn't include yet implementation of `hwmon_ina219`.  First, I've tried to compile standalone driver. It kind of worked, I could read the data from UPS but NUT in many places showed that there is no driver connected. After this, I've tried to make backport but I didn't succeed. I think it's because of differences between some libraries in `testing` and `stable` versions of Debian.

Since I don't have any additional Raspberry Pi laying around to test it fully on Debian `testing` or any other OS, I just gave up.

**Comment by jviki** [comment]
Yes, that issue with NUT at version 2.8.1 on Debian bothers me as well...

**Comment by mpn01** [comment]
The workaround I thought about is to enable device as hwmon, create docker container with attached volume to `/sys/class/hwmon/...` and then use it as a NUT server. It's not ideal but I think it should work.

**Comment by jviki** [comment]
Yes, that should work, good idea.
